File: Assets/ControlCanvas/SkillButton.js
#pragma strict
import UnityEngine.UI;
var skillIndex: int = 0;
var skillCaster: GameObject;
private var eventCenter: EventCenter;
private var skillProperties: SkillProperties;
private var coolDownTimer: CoolDownTimer;
private var coolDownTimeText: Text;
private var coolDownIndicatorRectTransform: RectTransform;

function Awake () {
	eventCenter = GameObject.FindWithTag('EventCenter').GetComponent(EventCenter);
	var playerSkillCasters: GameObject = GameObject.FindWithTag('PlayerSkillCasters');
	if(playerSkillCasters.transform.childCount > skillIndex) {
		skillCaster = playerSkillCasters.transform.GetChild(skillIndex).gameObject;
		skillProperties = skillCaster.GetComponent(SkillProperties);
		coolDownTimer = skillCaster.GetComponent(CoolDownTimer);
	}
	if(null != skillProperties) {
		GetComponentInChildren(Text).text = skillProperties.skillName;
		name = skillProperties.skillName + ' Button';
		GetComponent(Image).color = skillProperties.buttonColor;
	}
	if(null != coolDownTimer) {
		coolDownTimeText = GetComponentsInChildren(Text)[1] as Text;
		coolDownIndicatorRectTransform = (GetComponentsInChildren(Image)[1] as Image).rectTransform;
	}
}
function Update () {
	if(null != coolDownTimeText) {
		if(false == coolDownTimer.GetIsCoolDownFinished()) {
			var remainCoolDownTime: double = coolDownTimer.GetRemainCoolDownTime();
			if(1.0 < remainCoolDownTime) {
				coolDownTimeText.text = remainCoolDownTime.ToString('#');
			} else {
				coolDownTimeText.text = remainCoolDownTime.ToString('.#');
			}
			var coolDownTime: double = coolDownTimer.coolDownTime;
			coolDownIndicatorRectTransform.anchorMin.x = (1 - 1000*remainCoolDownTime/coolDownTime);
		} else {
			coolDownTimeText.text = '';
		}
	}
}
function OnPointerDown() {
	eventCenter.CastEvent(this, 'skillButtonDown', skillIndex);
}
function OnPointerUp() {
	eventCenter.CastEvent(this, 'skillButtonUp', skillIndex);
}

File: Assets/Script/Button/SkillButton.cs
using UnityEngine;
using UnityEngine.UI;

public class SkillButton : MonoBehaviour {
	public int skillIndex = 0;
	public SkillCasterBase skillCasterBase;
	public Timer coolDownTimer;
	public Text coolDownTimeText;
	public RectTransform coolDownIndicatorRectTransform;

	public void Awake () {
        EventManager.Instance.RegisterListener(EventManager.Instance, "casterReady", gameObject, OnCasterReady);
	}
	public void Update () {
		if(null != coolDownTimeText) {
			if((coolDownTimer != null) && coolDownTimer.isTiming) {
				float remainCoolDownTime = (float)coolDownTimer.GetRemainTime();
				if(1.0 < remainCoolDownTime) {
					coolDownTimeText.text = remainCoolDownTime.ToString("#");
				} else {
					coolDownTimeText.text = remainCoolDownTime.ToString(".#");
				}
				float coolDownTime = (float)coolDownTimer.duration;
				Vector2 anchorMin = coolDownIndicatorRectTransform.anchorMin;
				anchorMin.x = (1 - remainCoolDownTime/coolDownTime);
				coolDownIndicatorRectTransform.anchorMin = anchorMin;
			} else {
				coolDownTimeText.text = "";
			}
		}
	}
    public void OnCasterReady(SbiEvent e) {
        CasterReadyEventData data = (CasterReadyEventData)e.data;
		if(GameManager.Instance.GetPlayer() != data.player) { return; }
		if(skillIndex == data.skillIndex) {
			skillCasterBase = data.skillCaster.GetComponent<SkillCasterBase>();
            coolDownTimer = skillCasterBase.GetTimerByType("Cooldown");
		}
		if(null != skillCasterBase) {
            Debug.Log("OnCasterReady color:" + skillCasterBase.buttonColor.ToString() );
            GetComponentInChildren<Text>().text = skillCasterBase.skillName;
			name = skillCasterBase.skillName + " Button";
			GetComponent<Image>().color = skillCasterBase.buttonColor;
		}
		if(null != coolDownTimer) {
			coolDownTimeText = GetComponentsInChildren<Text>()[1] as Text;
			coolDownIndicatorRectTransform = (GetComponentsInChildren<Image>()[1] as Image).rectTransform;
		}
    }
	public void OnPointerDown() {
		EventManager.Instance.CastEvent(
			this, "skillButtonDown", new SkillButtonEventData("down", skillIndex)
		);
	}
	public void OnPointerUp() {
		EventManager.Instance.CastEvent(
			this, "skillButtonUp", new SkillButtonEventData("up", skillIndex)
		);
	}
}
